feat: preparation to add Unity->DCC Tools communication

schinkowski opened the pull request:
Adding 2 way communication between unity and blender.

Commit:
Additional checks and handling for missing components or materials

Commit:
Exposing geometry nodes parameters in the Instance renderer inspector

Commit:
Using Utility function for destroying components to avoid branching

Commit:
using clamp instead of min of max

Commit:
Instances stored in a dictionary for more efficient lookup

Commit:
Removed toggle button for activating instance rendering from config, instance renderer is a private field of the server instead of a monobehavior, instance rendering is always active

Commit:
Adding a way to run blender from inside unity

Commit:
Updates to blender run modes.

Commit:
Merge branch 'exposed-geometry-nodes-parameters-support-blender' of https://github.com/unity3d-jp/MeshSync into dublin_team

# Conflicts:
#	Editor/Scripts/MeshSyncServerInspector.cs
#	Runtime/Scripts/MeshSyncServer.cs

Commit:
Moving sliders to mesh sync server

Commit:
Merge remote-tracking branch 'origin/instance-rendering-support' into dublin_team

# Conflicts:
#	Runtime/Scripts/MeshSyncInstanceRenderer.cs
#	Runtime/Scripts/MeshSyncServer.cs
#	Runtime/Scripts/Settings/Data/MeshSyncPlayerConfig.cs

Commit:
prep

Commit:
Scene exporter uses the instance infos manager. msCore API adapted to use the instance infos in set messages. delete message support is WIP. Instance Renderer updated to listen to new events. Player Record class created for instance info, also WIP

Commit:
Updated how delete message is handled for instance infos

Commit:
Associating instances with entities instead of looking for gameobject in Scene

Commit:
WIP blender UI

Commit:
Support for receiving instances of meshes that are not part of the scene

Commit:
Merge remote-tracking branch 'origin/instance-rendering-support' into dublin_team

# Conflicts:
#	Runtime/Scripts/BaseMeshSync.cs
#	Runtime/Scripts/MeshSyncInstanceRenderer.cs

Commit:
Skip sending only if there are no instances or instance meshes

Commit:
Merge remote-tracking branch 'origin/instance-rendering-support' into dublin_team

# Conflicts:
#	Plugin~/Src/MeshSync/Include/MeshSync/SceneExporter.h
#	Plugin~/Src/MeshSync/Include/MeshSync/SceneGraph/msScene.h
#	Plugin~/Src/MeshSync/SceneGraph/msScene.cpp
#	Plugin~/Src/mscore/msCoreAPI.cpp
#	Runtime/Plugins/x86_64/mscore.dll
#	Runtime/Scripts/msCoreAPI.cs

Commit:
WIP property support

Commit:
Fix for instances not rendering on 2-window workflow

Commit:
Improved Fix for out of focus instance rendering

Commit:
Single camera support for Play mode

Commit:
Mutliple gameview camera support for Playmode

Commit:
WIP properties

Commit:
Support for multiple game views or multiple scene views rendering in edit mode

Commit:
Support for chaging camera mode from inspector. Instanting Rendering parameters are reflected from original gameobjects

Commit:
Support for moving/rotating/scaling instances when the referenced gameobject or a parent is moving/rotating/scaling

Commit:
Support for moving/rotating/scaling instances when the referenced gameobject or a parent is moving/rotating/scaling

Commit:
WIP properties

Commit:
Only redraw at the end of the set message sequence

Commit:
Support for light probe proxy volume

Commit:
Caching for instance world matrix recalculation

Commit:
Support for IL2CPP builds

Commit:
WIP properties

Commit:
Merge remote-tracking branch 'origin/instance-rendering-support' into dublin_team

# Conflicts:
#	Editor/Scripts/MeshSyncServerInspector.cs
#	Runtime/Scripts/BaseMeshSync.cs

Commit:
Serialization support for instancing info

Commit:
Pin data to avoid issues with memory defragmentation

Commit:
Removed instance renderer clear from init

Commit:
Render to all cameras by default

Commit:
Fix for moveability error

Commit:
Try to reduce memory pressure at the end of a scene message

Commit:
Reverted manual garbage collection

Commit:
Changed c++ version to c++17

Commit:
Support to control geo node properties from unity.

Commit:
C++17 flag set for Apple

Commit:
Fix for race condition

Commit:
Avoid compilation errors when an unsupported build target is selected

Commit:
Handled merge conflicts with develop

Commit:
Fix for rendering issue when Undo/Redo is invoked

Commit:
Fix for compilation issues

Commit:
Handled merge conflicts with develop

Commit:
Merge https://github.com/unity3d-jp/MeshSync into instance-rendering-support

Commit:
Handled merge conflicts

Commit:
Implementing custom properties

Commit:
Fix: Support for instances of SkinnedMeshRenderer gameobjects

Commit:
Updated linux build instructions about using c++17 dev tools instead of c++14

Commit:
WIP perrmutation baker

Commit:
Fix: instance rendering handling for records without gameobject info

Commit:
Implemented copy and move ctors and assignment operators

Commit:
Avoid rendering instances while the scene is updating

Commit:
Merge remote-tracking branch 'origin/instance-rendering-support' into dublin_team

# Conflicts:
#	Runtime/Plugins/x86_64/mscore.dll
#	Runtime/Scripts/MeshSyncInstanceRenderer.cs
#	Runtime/Scripts/MeshSyncServer.cs

Commit:
Merge remote-tracking branch 'origin/instance-rendering-support' into dublin_team

Commit:
Geometry Nodes Support: Refactored how instanced meshes are exported, managed and sent. All instancing info is now sent in one frame. Removed need for reference type in geometry nodes

Commit:
Refactor: Remove Reference type from instance info. Simplified instance deletion code

Commit:
Refactor: simplified instance rendering by using the RenderPipelineManager.beginFrameRendering and Camera.OnPreCull events

Commit:
Instance Rendering: avoid rendering on the Preview Scene Camera on Built-in pipeline

Commit:
Handled merge conflicts with dev

Commit:
WIP variant export

Commit:
Refactor: Instance rendering and instance serialization. Each instanced gameobject has its own rendering component. The transform data are serialized in the gameobject

Commit:
Instance Rendering: Support for saving depending material and meshes in the Asset database when creating a prefab of an instance renderer

Commit:
wip baking

Commit:
Merge remote-tracking branch 'origin/instance-rendering-support' into dublin_team

# Conflicts:
#	Plugin~/Src/MeshSync/AsyncSceneSender.cpp
#	Plugin~/Src/MeshSync/Include/MeshSync/SceneGraph/msScene.h
#	Runtime/Plugins/x86_64/mscore.dll
#	Runtime/Scripts/MeshSyncInstanceRenderer.cs
#	Runtime/Scripts/MeshSyncServer.cs
#	Runtime/Scripts/PlayerData/InstanceInfoRecord.cs

Commit:
Instance Rendering: changed layout of instance rendering components to match blender3d

Commit:
refactoring

Commit:
Merge remote-tracking branch 'origin/instance-rendering-support' into dublin_team

Commit:
Instances: Improved the records and scene management for instance rendering components

Commit:
Instance Rendering: Allow instance rendering while scene is updating to avoid flickering

Commit:
merge fix

Commit:
Merge remote-tracking branch 'origin/instance-rendering-support' into dublin_team

# Conflicts:
#	Runtime/Scripts/msCoreAPI.cs

Commit:
Fix for render pipeline calllback

Commit:
Refactoring and fixes

Commit:
Instance Rendering: Use GraphicsSettings to check for the current rendering pipeline

Commit:
Instance Rendering: use root object as parent if parent is missing

Commit:
Instance Rendering: Fix for server poll delegate not unregistering from EditorApplication update. Hiding transforms from Inspector for avoiding GUI rendering performance issues

Commit:
Instance Rendering: Serialize inverse rendering info components to avoid issues with recompilation

Commit:
Serialization: Fixed Asset modified processor to handle all prefabs, not just instance rendering references

Commit:
Fixes and improvements for 2 way communication and baking

Commit:
Adding new exporter

Commit:
Instance Rendering: avoid rendering instanced on prefab scenes that do not contain the instanced gameobject

Commit:
Merge remote-tracking branch 'origin/instance-rendering-support' into dublin_team

# Conflicts:
#	Runtime/Scripts/MeshSyncInstanceRenderer.cs

Commit:
Instance Rendering: Fixed unity runtime building errors. Better re-use of the transform/mesh entity update functions

Commit:
Fixes and refactoring for 2 way communication

Commit:
wip

Commit:
Merge remote-tracking branch 'origin/instance-rendering-support' into dublin_team

# Conflicts:
#	Runtime/Scripts/MeshSyncInstanceRenderer.cs

Commit:
Fix for property request not timing out

Commit:
Refactoring and bug fixes

Commit:
Improved asset name sort

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/unity3d-jp/MeshSync?pullRequest=639) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/unity3d-jp/MeshSync?pullRequest=639) before we can accept your contribution.<br/><hr/>**Christian Schinkoethe** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/unity3d-jp/MeshSync?pullRequest=639) it.</sub>

Commit:
Fix to save changed meshes for prefab variants

Commit:
More property support

Commit:
Property fixes

Commit:
Fix for running meshsync in background mode

Commit:
UI Improvements and fixes

Commit:
Documentation for geometry nodes [skip ci]

Commit:
geomentry node docs [skip ci]

Commit:
Update Geometry Nodes.md [skip ci]

Commit:
Added bakemodstups image

Commit:
Merge branch 'geometry-nodes-docs' of https://github.com/unity3d-jp/MeshSync into geometry-nodes-docs

Commit:
Added instance light probes gif

Commit:
Updated GeometryNodes.md [skip ci]

Commit:
WIP curve support

Commit:
Use separate script to store blender properties

Commit:
Merge remote-tracking branch 'origin/instances-light-probes-docs' into dublin_team

# Conflicts:
#	Runtime/Scripts/BaseMeshSync.cs

Commit:
Fixes for prefab export. Implemented different instance handling approaches.

Commit:
Cleanup

Commit:
Ensure visibility is updated from autosync.

Commit:
Fix for asset expoort and sharing assets between prefabs

Commit:
Fix for clashing property names. Refactoring

Commit:
Improved UI

Commit:
Adding curve support

Commit:
Curve fixes and multiple spline support.

Commit:
WIP curve refactor and mesh support.

Commit:
WIP mesh support.

Commit:
module dependency fixes

Commit:
Adding event to call before export.

Commit:
Limit regenerator to controller.
Fix for string properties.

Commit:
Added mesh support

Commit:
clean up

Commit:
fix for crash and refactor

Commit:
Adding dependencies

Commit:
Disable quads to make it faster

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into dublin_team

# Conflicts:
#	Documentation~/GeometryNodes.md
#	MeshSync_URP~/Packages/manifest.json
#	MeshSync_URP~/Packages/packages-lock.json
#	MeshSync_URP~/ProjectSettings/ProjectVersion.txt
#	Plugin~/Src/MeshSync/AsyncSceneSender.cpp
#	Plugin~/Src/MeshSync/Include/MeshSync/SceneExporter.h
#	Plugin~/Src/MeshSync/Include/MeshSync/SceneGraph/msInstanceInfo.h
#	Plugin~/Src/MeshSync/Include/MeshSync/SceneGraph/msScene.h
#	Plugin~/Src/MeshSync/SceneGraph/msInstanceInfo.cpp
#	Plugin~/Src/MeshSync/SceneGraph/msScene.cpp
#	Plugin~/Src/mscore/msCoreAPI.cpp
#	Runtime/Plugins/x86_64/mscore.bundle/Contents/MacOS/mscore
#	Runtime/Plugins/x86_64/mscore.dll
#	Runtime/Scripts/BaseMeshSync.cs
#	Runtime/Scripts/InstanceRenderingInfo.cs
#	Runtime/Scripts/MeshSyncInstanceRenderer.cs
#	Runtime/Scripts/MeshSyncServer.cs
#	Runtime/Scripts/PlayerData/InstanceInfoRecord.cs
#	Runtime/Scripts/msCoreAPI.cs

Commit:
Merge fixes

Commit:
Adding validation to ensure assets are blend files

Commit:
Removing not needed stuff

Commit:
Removing not needed stuff

Commit:
Cleanup

Commit:
Fix for serialization

Commit:
Removing unused setting

Commit:
Removing variant exporter

Commit:
Clean up

Commit:
Formatting fix

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into dublin_team

Commit:
Updating comment

Commit:
Removing not needed stuff.

Commit:
Updating compiled dll

sindharta commented on Runtime/Unity.MeshSync.asmdef:
```diff
@@ -35,6 +39,16 @@
+        {
+            "name": "com.unity.splines",
+            "expression": "2.0.0-pre",
+            "define": "MESHSYNC_SPLINE_SUPPORT"
```
The convention that we use for defines is `AT_USE_FOO`

Could you please update this ?

schinkowski replied:
Fixed

sindharta commented on Runtime/Scripts/msCoreAPI.cs:
```diff
@@ -2986,9 +3142,15 @@ internal struct SceneData {
+#endregion
 
-    public static implicit operator bool(SceneData v) {
+        public static implicit operator bool(SceneData v) {
```
Why are there some tab/spacing changes occasionally, which actually make the code not lined up ?

schinkowski replied:
My VS is set up to auto-format the code, the whole file has a different formatting convention so when VS reformats this line it's indented too far. After this merge we should format the entire project to use a consistent formatting style.

sindharta commented on Runtime/Scripts/msCoreAPI.cs:
```diff
@@ -2945,8 +3034,75 @@ public Matrix4x4[] transforms
+
+#region PropertyInfo
+
+    public struct PropertyInfoData
```
Please make this internal

schinkowski replied:
I need this to be public because of the regenerator we are building for Bay. This is built on top of MeshSync as a separate tool and requires access to the blender properties to change them and get updated meshes out of blender.

schinkowski replied:
Actually, I've changed it now and made the enums inside it public, that's all I really need.

sindharta commented on Runtime/Scripts/msCoreAPI.cs:
```diff
@@ -2736,15 +2756,84 @@ internal struct PointsData {
-internal struct ConstraintData {
-    #region internal
+    internal struct ConstraintData {
+#region internal
```
I don't mind this change of code convention for `#region` so that it's not preceded by spaces, but it would be great to have this change in a separate PR.

schinkowski replied:
Fixed

sindharta commented on Runtime/Scripts/msCoreAPI.cs:
```diff
@@ -22,12 +26,17 @@ internal static class Lib {
+    static string version;
+    internal static string GetPluginVersion()
+    {
+        if (version == null)
```
I think this is an optimization change which is unrelated to the new functionality.
Would be great to have this separated in another PR.

schinkowski replied:
Done here: https://github.com/unity3d-jp/MeshSync/pull/696

sindharta commented on Editor/Unity.MeshSync.Editor.asmdef:
```diff
@@ -30,6 +35,16 @@
+        {
+            "name": "com.unity.splines",
+            "expression": "2.0.0-pre",
+            "define": "MESHSYNC_SPLINE_SUPPORT"
```
The convention that we use for defines is `AT_USE_FOO`

Could you please update this ?

schinkowski replied:
Fixed

sindharta commented on MeshSync_URP~/Packages/manifest.json:
```diff
@@ -1,13 +1,15 @@
-    "com.unity.render-pipelines.universal": "7.7.1",
+    "com.unity.probuilder": "5.0.4",
+    "com.unity.render-pipelines.universal": "12.1.3",
+    "com.unity.splines": "git@github.cds.internal.unity3d.com:unity/com.unity.splines.git",
```
This requires access to an internal repo. I would suggest creating a new project internally for this purpose.

schinkowski replied:
This is an optional dependency and only set in this test project for debugging. I'm checking with the splines team if there is a public github repo for this yet. I don't want to fork off their repo and make it public if they didn't officially release it.

sindharta commented on Plugin~/Src/MeshSync/Include/MeshSync/SceneGraph/msEntityConverter.h:
```diff
@@ -1,20 +1,18 @@
 
 namespace ms {
 
-class Entity;
```
We used forward declarations here to reduce dependencies in header files directly.
Is it necessary to include  all the header files ?

schinkowski replied:
Fixed

sindharta commented on Plugin~/Src/MeshSync/Include/MeshSync/msServer.h:
```diff
@@ -122,8 +131,31 @@ class Server {
+    std::vector<EntityPtr> m_pending_entities;
+
+    template<typename T>
+    T* getPendingEntity(const char* path) {
```
please rename to `getOrCreate~`

sindharta replied:
because just `get~` implies that it only gets, which should be cheap and does not modify internal member variable

sindharta replied:
Bumping up

sindharta commented on Plugin~/Src/MeshSync/msClient.cpp:
```diff
@@ -168,6 +171,71 @@ bool Client::send(const FenceMessage& mes)
+
+        m_properties_session = &session;
+
+        session.setTimeout(Poco::Timespan()); // infinite timeout, this is cancelled manually.
```
Why do we have infinite timeout ?

schinkowski replied:
Because this request is kept alive until needs to send something back.

sindharta commented on Plugin~/Src/MeshSync/msClient.cpp:
```diff
@@ -168,6 +171,71 @@ bool Client::send(const FenceMessage& mes)
+    }
+}
+
+bool Client::send(const ServerInitiatedMessage& mes)
```
Trying to clarify.
So the client sends `ServerInitiatedMessage` , and the server responds by sending `ServerInitiatedMessageResponse` to the client ?

If so, then is `ServerInitiatedMessage` not misleading because it's not initiated by the server ?

schinkowski replied:
The server doesn't have a way to request or send data to the client directly. That's why the client sends this request that the server can respond to when it does want to send something. 
What would be a better name for it?

sindharta replied:
I understand more about the flow after reading the documentation on Confluence.

Perhaps naming it explicitly to what it does is better ? Something like `ServerLiveEditRequest` and `ServerLiveEditResponse`

schinkowski replied:
Done

sindharta commented on Plugin~/Src/MeshSync/AsyncSceneSender.cpp:
```diff
@@ -66,6 +71,48 @@ void AsyncSceneSender::kick()
     m_future = std::async(std::launch::async, [this]() { send(); });
 }
 
+void AsyncSceneSender::requestServerInitiatedMessage()
```
Where is this called ?

schinkowski replied:
This is called by msblenContext from the update loop in python.

sindharta commented on Runtime/Scripts/msNetworkAPI.cs:
```diff
@@ -145,6 +187,66 @@ internal struct Server {
+        msServerRequestFullSync(self);
+    }
+
+    public void SendChangedProperties()
```
It seems that this function will change the `ready` flag of `m_current_properties_request` which is required by `Server::recvServerInitiatedRequest` ?

If so, perhaps we should rename the function as it's not actually sending changed properties ?

schinkowski replied:
Ok, I've renamed it to better reflect what it does.

sindharta commented on Editor/Scripts/MeshSyncServerPropertiesInspector.cs:
```diff
@@ -0,0 +1,279 @@
+
+namespace Unity.MeshSync.Editor
+{
+    public static class MeshSyncServerInspectorUtils
```
Can we put this class in another file  ?

sindharta commented on Editor/Scripts/DCCIntegration/Launchers/BlenderLauncher.cs:
```diff
@@ -0,0 +1,253 @@
+
+            if (string.IsNullOrEmpty(blenderPath))
+            {
+                if (!DefaultAppUtility.TryGetRegisteredApplication(".blend", out blenderPath))
```
Does this work on Windows only ? If so, we need to wrap this inside `#if UNITY_EDITOR_WIN`

schinkowski replied:
Yeah, fixed

sindharta commented on Editor/Scripts/DCCIntegration/Launchers/BlenderLauncher.cs:
```diff
@@ -0,0 +1,253 @@
+        }
+
+        [MenuItem(menuItem_OpenInBlender, priority = 0)]
+        static void OpenBlender()
```
Can we separate changes that add functionality to open a selected object in Blender into a different PR ?

schinkowski replied:
Done here: https://github.com/unity3d-jp/MeshSync/pull/697

sindharta commented on Runtime/Scripts/MeshSyncServer.cs:
```diff
@@ -455,14 +478,24 @@ public void StartServer()
+        if (m_DCCAsset != null)
+        {
+            var path = AssetDatabase.GetAssetPath(m_DCCAsset);
+            if (System.IO.Path.GetExtension(path) != ".blend")
```
Are we assuming that we only handle Blender ?
This is not the case as I am sure you are already aware of.

schinkowski replied:
Yeah good point, I've refactored this so it doesn't check the filetype anymore.

sindharta commented on Editor/Scripts/MeshSyncServerInspector.cs:
```diff
@@ -1,12 +1,81 @@
+    static MeshSyncServerInspector()
+    {
+        BaseMeshSync.ProBuilderBeforeRebuild += ProBuilderBeforeRebuild;
+        BaseMeshSync.ProBuilderAfterRebuild += ProBuilderAfterRebuild;
```
New Callbacks will be registered everytime we show the inspector of MeshSyncServer.
Is this expected ?

schinkowski replied:
This is the static constructor, it's only called once and it has the [InitializeOnLoad] attribute so it's also registered before it shows.

sindharta commented on Editor/Scripts/MeshSyncServerInspector.cs:
```diff
@@ -85,6 +155,44 @@ public void DrawServerSettings(MeshSyncServer t)
+
+                EditorGUILayout.LabelField(prefabHolder.name);
+
+                EditorGUILayout.ObjectField(prefabHolder.prefab, typeof(GameObject), true);
```
there is a `EditorGUILayout.ObjectField()` function which receives a label, if that's what is intended here.

schinkowski replied:
Thanks, fixed

sindharta commented on Plugin~/Src/MeshSync/SceneGraph/msScene.cpp:
```diff
@@ -231,27 +233,62 @@ void Scene::sanitizeObjectName(std::string& name)
+        else if (cv.zup_correction_mode == ZUpCorrectionMode::RotateX) {
+            if (invert) {
+                // TODO: Could write a new RotateX_ZUpCorrector that rotates back, for now just rotate 3 more times to get back to original:
+                for (int i = 0; i < 3; i++)
```
I don't understand why doing the same rotation three times would be the same as doing the inverse rotation. Is this the case ?

schinkowski replied:
Yeah, the converter has a hardcoded rotation of 90 degrees, i just rotate 3 more times at 90 degrees to get back where it was. After this is merged I'm going to refactor that to set the rotation in its constructor or something.

sindharta replied:
It does some swapping and flipping while doing the rotation, and the transformation isn't the same for position, rotation and scale.
Are you sure that doing this 3 times will invert all of them to where they were ?
```
    if (e.isRoot()) {
        e.position = flip_z(swap_yz(e.position));
        e.rotation = flip_z(swap_yz(e.rotation)) * mu::rotate_x(-90.0f * mu::DegToRad);
        e.scale = swap_yz(e.scale);
    }
```

Might be worthy to turn this into a unit test.

schinkowski replied:
I've added unit tests to make sure this works. Turns out the RotateX_ZUpCorrector needs 8 transformations because of the z flipping. I hadn't used that much because it's buggy anyway.

Commit:
Adding mathematics package dependency.
Rebuilt dll for release.

Commit:
Updating protocol version.
Extracted string constant.

Commit:
Fix for prefab's asset dependency not being deleted.

Commit:
Refactoring

Commit:
Refactoring

Commit:
Renaming define

Commit:
Formatting revert

Commit:
Removing unused code

Commit:
Formatting revert

Commit:
Refactoring

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into dublin_team

# Conflicts:
#	Runtime/Scripts/MeshSyncServer.cs

Commit:
Adding spaces to fix macro string concatenation.

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into dublin_team

Commit:
Fix for dependencies and packages to work with unity 2019.

Commit:
Fix for windows only functionality

Commit:
Refactoring to draw ui for dcc tools in the meshsyncserver inspector.

Commit:
Using version compatible with unity 2019

Commit:
rebuilt dll

Commit:
Refactoring

Commit:
Fix for string concatenation

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into dublin_team

# Conflicts:
#	Runtime/Plugins/x86_64/mscore.dll

Commit:
Updating unit tests.

Commit:
Updated dlll

Commit:
Updated tests.

Commit:
Fix for latest splines package code changes.

Commit:
Reverting unwanted change

Commit:
revert

Commit:
Reverting not needed change

schinkowski commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -70,674 +74,668 @@ namespace Unity.MeshSync
 /// </summary>
 internal delegate void DeleteInstanceHandler(string path);
```
separate PR for this reformat here: https://github.com/unity3d-jp/MeshSync/pull/701

Commit:
Renaming method to be more clear what it does.

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into dublin_team

# Conflicts:
#	Runtime/Plugins/x86_64/mscore.dll
#	Runtime/Scripts/BaseMeshSync.cs
#	Runtime/Scripts/MeshSyncServer.cs

Commit:
Removing script that causes issues in the build pipeline.

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into dublin_team

Commit:
Compile fixes when building game.

Commit:
reverting changes

Commit:
Reverting changes

Commit:
Reverting changes

Commit:
Format revert

Commit:
Format fix

Commit:
Updating protocol version

Commit:
Fix for compiler errors on mac and linux

sindharta committed:
Update linux so file
[skip ci]

sindharta-tanuwijaya committed:
update Mac intel bundle
[skip ci]

sindharta-tanuwijaya committed:
Update mscore mac Silicon bundle

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into dublin_team

# Conflicts:
#	Editor/Scripts/DCCIntegration/Launchers/BlenderLauncher.cs
#	Editor/Scripts/MeshSyncMenu.cs
#	Editor/Scripts/MeshSyncServerInspector.cs
#	Editor/Scripts/MeshSyncServerInspectorUtils.cs
#	Runtime/Scripts/MeshSyncServer.cs

seandillon92 commented on Editor/Scripts/MeshSyncServerInspector.cs:
```diff
@@ -1,12 +1,81 @@
+
+    private static void MeshesChanged(IEnumerable<UnityEngine.ProBuilder.ProBuilderMesh> meshes)
+    {
+        if (meshes != null)
```
can this be inverted?

schinkowski replied:
Done

Commit:
Deleting not needed code.

seandillon92 commented on Editor/Scripts/MeshSyncServerInspector.cs:
```diff
@@ -1,12 +1,81 @@
+    private static void ProBuilderBeforeRebuild()
+    {
+        // Change select mode back and forth to ensure probuilder invalidates its internal cache:
+        var t = UnityEditor.ProBuilder.ProBuilderEditor.selectMode;
```
can this be renamed to something like lastSelectMode?

schinkowski replied:
Done

seandillon92 commented on Editor/Scripts/MeshSyncServerInspector.cs:
```diff
@@ -1,12 +1,81 @@
+    {
+        // Change select mode back and forth to ensure probuilder invalidates its internal cache:
+        var t = UnityEditor.ProBuilder.ProBuilderEditor.selectMode;
+        UnityEditor.ProBuilder.ProBuilderEditor.selectMode = UnityEngine.ProBuilder.SelectMode.Object;
+        UnityEditor.ProBuilder.ProBuilderEditor.selectMode = t;
```
Will this work if t is SelectMode.Object?

schinkowski replied:
Yes but then the mode would change for the user when they don't want to. I just need to change the mode because ProBuilder doesn't support an API to refresh the state when the underlying data changed.

seandillon92 replied:
I don't suggest we change t to SelectMode.Object in the expression. I am not sure if this will work if the value of t is Object. Will ProBuilder refresh in this case?

seandillon92 commented on Plugin~/Src/MeshSync/Include/MeshSync/AsyncSceneSender.h:
```diff
@@ -25,15 +28,19 @@ class AsyncSceneSender : public SceneExporter
     std::string m_error_message;
-};
-
+    std::atomic_bool destroyed{ false };
```
Should this be named m_destroyed? to follow the convention in the file

schinkowski replied:
Done

seandillon92 commented on Plugin~/Src/MeshSync/msServerRequestHandler.cpp:
```diff
@@ -80,6 +80,9 @@ void ServerRequestHandler::handleRequest(HTTPServerRequest& request, HTTPServerR
     else if (StartsWith(uri, "/plugin_version")) {
         m_server->serveText(response, msPluginVersionStr);
     }
+    else if (StartsWith(uri, "request_properties")) {
```
why is there no "/"? seems that all the uris above use it?

schinkowski replied:
Some use it, some don't. I've added it now.

seandillon92 commented on Plugin~/Src/mscore/msServerAPI.cpp:
```diff
@@ -247,6 +248,120 @@ msAPI void msServerNotifyPoll(ms::Server *server, ms::PollMessage::PollType t)
+
+    server->m_pending_entities.push_back(make_shared_ptr(data));
+
+    //auto mesh = server->getPendingEntity<ms::Mesh>(path);
+
+    //mesh->points.assign(vertices, vertices + vertexCount);
+
+    //// Update flags:
+    //mesh->refine();
```
dead code?

schinkowski replied:
Fixed

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Modifiers.cs:
```diff
@@ -0,0 +1,552 @@
+            Try(() =>
+            {
+                var numProperties = scene.numPropertyInfos;
+                if (numProperties > 0)
```
can we invert this?

schinkowski replied:
Fixed

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Modifiers.cs:
```diff
@@ -0,0 +1,552 @@
+
+                        propertyInfos.Sort(new PropertyInfoDataWrapperComparer());
+
+                        if (pendingProps != null)
```
can this be inverted as well?

schinkowski replied:
Fixed

seandillon92 commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1775,46 +1786,173 @@ EntityRecord UpdateInstancedEntity(TransformData data)
+                    infoRecord.renderer.UpdateAll(data.transforms, infoRecord.go);
+                    break;
+                case InstanceHandlingType.Copies:
+                case InstanceHandlingType.Prefabs:
```
This switch case is really long. It also seems to be having some checks on the variable, which makes the code hard to follow. Might be better if the code was broken down into smaller functions and the Copies had its own function that reuses the common stuff from the prefabs.

As a first step, it would be great if some of the statement was broken down into smaller functions.

schinkowski replied:
Done

seandillon92 commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,308 @@
+            }
+        }
+
+        void SendTransform(string path, EntityRecord entity)
+        {
+
+        }
```
dead code?

schinkowski replied:
It's something to implement later. I've created a ticket for it here: https://jira.unity3d.com/browse/BLENDER-478

seandillon92 commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,308 @@
+            //}
+            //else if (objRenderer.GetType() == typeof(SkinnedMeshRenderer)){
+            //    ret = CaptureSkinnedMeshRenderer(ref dst, objRenderer as SkinnedMeshRenderer, mes, ref origMesh);
+            //}
```
Might be better we raise a jira ticket for this and remove the dead code?

schinkowski replied:
I created a ticket for it here: https://jira.unity3d.com/browse/BLENDER-477

seandillon92 commented on Runtime/Scripts/MeshSyncServer.cs:
```diff
@@ -14,16 +15,26 @@ namespace Unity.MeshSync {
     
 //----------------------------------------------------------------------------------------------------------------------
-    
+
+#if ENABLE_IL2CPP
+    // When building with IL2CPP, we must use a static delegate
+    private static MeshSyncServer Instance;
+
+    [MonoPInvokeCallback(typeof(Server.MessageHandler))]
+    private static void HandleReceivedMessage(NetworkMessageType type, IntPtr data)
+    {
+        Instance.HandleRecvMessage(type, data);
+    }
+#endif
```
Please remove this. I wrote this originally, but we can create a separate PR to tackle the IL2CPP builds

schinkowski replied:
Done

seandillon92 commented on Runtime/Scripts/MeshSyncServer.cs:
```diff
@@ -107,9 +118,14 @@ public void StartServer()
         m_server.fileRootPath = GetServerDocRootPath();
         m_server.AllowPublicAccess(projectSettings.GetServerPublicAccess());
         
+#if ENABLE_IL2CPP
+        m_handler = HandleReceivedMessage;
+#else
```
Please remove this too

schinkowski replied:
Done

seandillon92 commented on Runtime/Scripts/MeshSyncServer.cs:
```diff
@@ -485,16 +509,63 @@ public void StartServer()
-    void LateUpdate() {
+    void Start()
+    {
+#if ENABLE_IL2CPP
+        Instance = this;
+#endif
```
Please remove this too

schinkowski replied:
Done

seandillon92 commented on Runtime/Scripts/PluginAPI/NetworkData/GetFlags.cs:
```diff
@@ -1,21 +1,47 @@
+        public bool GetUV(int index) { return m_flags[UV_START_BIT_POS + index]; }
+
+
+        public enum GetFlagsSetting
+        {
+            Transform = 0,
+            Points = 1,
+            Normals = 2,
+            Tangents = 3,
+            Colors = 6,
+            Indices = 7,
+            MaterialIDS = 8,
+            Bones = 9,
+            BlendShapes = 10
```
nice!

Commit:
Removing not needed code.

Commit:
Inverting if

Commit:
Refactoring

Commit:
Renaming field.

Commit:
Reducing included files.

Commit:
Fix for missing prefabs.
Renaming property request to match naming scheme of other requests of that kind.

Commit:
Removing dead code.

Commit:
Refactoring

Commit:
Refactoring

Commit:
Extracting methods.

Commit:
Adding TODO

Commit:
Removing commented out code

Commit:
Compile fix

sindharta commented on Plugin~/Src/MeshSyncTest/TestMeshSyncClient.cpp:
```diff
@@ -18,8 +19,101 @@
+
+TestCase(Test_ServerInitiatedMessage) {
+    ms::AsyncSceneSender sender;
+    sender.requestServerInitiatedMessage();
```
I tried to run this in debug mode, and then it could not continue past this code, probably due to the no timeout setting. I guess we can add some flags that will let the server send some mock data.
If necessary, these flags can also be turned on or defined only when `BUILD_TESTS` is defined in the project.

This particular code also logs this exception in my console:
```
Test_ServerInitiatedMessage begin
Exception thrown at 0x00007FFB37214FD9 in MeshSyncTest.exe: Microsoft C++ exception: Poco::Net::InvalidSocketException at memory location 0x000000834A9FF120.
```

sindharta replied:
It sometimes doesn't wait and logs these exceptions instead:
```
Test_ServerInitiatedMessage begin
Exception thrown at 0x00007FFB37214FD9 in MeshSyncTest.exe: Microsoft C++ exception: Poco::Net::InvalidSocketException at memory location 0x000000363B6FDAE0.
Exception thrown at 0x00007FFB37214FD9 in MeshSyncTest.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFB37214FD9 in MeshSyncTest.exe: Microsoft C++ exception: Poco::Net::InvalidSocketException at memory location 0x000000363B6FDAE0.
Exception thrown at 0x00007FFB37214FD9 in MeshSyncTest.exe: Microsoft C++ exception: Poco::Net::NetException at memory location 0x000000363B6FDFF0.
Exception thrown at 0x00007FFB37214FD9 in MeshSyncTest.exe: Microsoft C++ exception: Poco::Net::InvalidSocketException at memory location 0x000000363B6FBB60.
Test_ServerInitiatedMessage end (265.62ms)
```

I can still see the object being sent to Unity, so I have made sure that the server is active.

schinkowski replied:
It works withouut issues for me in debug mode, the destructor of the AsyncSceneSender aborts the network request, that gives me errors in the console because a thread is terminated by force.
The server is not guaranteed to be running and if Unity is running, it is not in test mode.
We'd need to set this up so it connects to a mock server in these test to properly test the client-server interaction.
Do we already have a BUILD_TESTS define? It doesn't seem to work for me, I don't see any custom defines in the test project at all.

sindharta replied:
> Do we already have a BUILD_TESTS define?

Hm, I think you are already doing this, but just in case, we can setup a test project by making the project using:

```
$ cmake -G "Visual Studio 15 2017" -A x64 .. -DBUILD_TESTS=ON
```

This is written in the [BuildPlugins](https://github.com/unity3d-jp/MeshSync/blob/dev/Plugin~/Docs/en/BuildPlugins.md#tips) page.

So to run the test:
1. Create the test solution
2. Build the test solution
3. Launch MeshSync~ test project and create MeshSyncServer gameobject
4. Run the MeshSyncTest project in the solution. 

Then you should see some objects appearing in Unity because the tests are sending them.
There are also other tests not directly related to Unity, like the new tests that you have written.

schinkowski replied:
I'm building it that way already. BUILD_TESTS is not defined.
I've increased the timeout, can you test if that fixes the tests getting stuck for you?
We should not need a different timeout for tests because the request is aborted.

sindharta replied:
It gives me these exception errors now:
```
Test_ServerInitiatedMessage begin
Exception thrown at 0x00007FFB37214FD9 in MeshSyncTest.exe: Microsoft C++ exception: Poco::IOException at memory location 0x0000008369AFD430.
Exception thrown at 0x00007FFB37214FD9 in MeshSyncTest.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFB37214FD9 in MeshSyncTest.exe: Microsoft C++ exception: Poco::IOException at memory location 0x0000008369AFD430.
Exception thrown at 0x00007FFB37214FD9 in MeshSyncTest.exe: Microsoft C++ exception: Poco::Net::NetException at memory location 0x0000008369AFE250.
Exception thrown at 0x00007FFB37214FD9 in MeshSyncTest.exe: Microsoft C++ exception: Poco::IOException at memory location 0x0000008369AFBDE0.
Test_ServerInitiatedMessage end (2015.90ms)
```

Do you have these exception errors on your side ?

sindharta commented on Editor/Scripts/MeshSyncServerInspector.cs:
```diff
@@ -1,12 +1,78 @@
+
+    static MeshSyncServerInspector()
+    {
+        BaseMeshSync.ProBuilderBeforeRebuild += ProBuilderBeforeRebuild;
```
Shouldn't we put the code to setup ProBuilder callbacks in another class ?
Why do we need to put it inside `MeshSyncServerInspector` ?

schinkowski replied:
We need a reference to the UnityEditor ProBuilder assemblies, these don't exist in the game assembly. This editor is guaranteed to exist when the mesh sync server is interacted with.

sindharta replied:
> these don't exist in the game assembly.

Right, but why can't we create another editor class which sets up these callbacks ?
The reason is that: 

1. this setup doesn't seem to be related to MeshSyncServerInspector directly (all the methods are static too)
2. creating a new class will keep the callback-setup code separated from MeshSyncServerInspector, which is already big.

schinkowski replied:
Done

sindharta commented on Plugin~/Src/MeshSync/msServer.cpp:
```diff
@@ -620,6 +681,38 @@ void Server::notifyPoll(PollMessage::PollType t)
+void Server::receivedProperty(PropertyInfoPtr prop) {
+    // If the same property is already prepared to be sent, replace it with the updated data:
+    for (size_t i = 0; i < m_pending_properties.size(); ++i) {
+        if (m_pending_properties[i]->matches(prop)) {
```
if we don't need a vector structure, perhaps we  can use an `std::map` using `msPropertyInfo::hash`, which allows us to do a direct lookup here.

schinkowski replied:
Fixed

sindharta commented on Runtime/Scripts/MeshSyncServerProperties.cs:
```diff
@@ -0,0 +1,33 @@
+    /// <summary>
+    /// Holds meta data about the settings that were used to create the object.
+    /// </summary>
+    internal class MeshSyncServerProperties : MonoBehaviour
```
Perhaps `MeshSyncServerLiveEditProperties` describes what this does better ?

schinkowski replied:
Fixed

sindharta commented on Runtime/Scripts/MeshSyncServerProperties.cs:
```diff
@@ -0,0 +1,33 @@
+
+        MeshSyncServer _server;
+
+        public MeshSyncServer Server
```
Doesn't seem to be used.
If it does, perhaps using [RequireComponent](https://docs.unity3d.com/ScriptReference/RequireComponent.html) is better ?

schinkowski replied:
Removed.

sindharta commented on Runtime/Scripts/MeshSyncServer.cs:
```diff
@@ -55,7 +55,7 @@ public class MeshSyncServer : BaseMeshSync {
 
 #if UNITY_STANDALONE        
-        if (m_autoStartServer && !m_serverStarted && enabled && gameObject.activeInHierarchy) {
+        if (m_autoStartServer && !m_serverStarted && gameObject.scene.IsValid() && !IsInPrefabView && enabled && gameObject.activeInHierarchy) {
```
This change has been merged to dev.
I wonder why we are seeing this here....

sindharta commented on Editor/Scripts/MeshSyncServerInspectorUtils.cs:
```diff
@@ -8,6 +8,231 @@ namespace Unity.MeshSync.Editor
+            {
+                case 2:
+                    {
+                        newValue = EditorGUILayout.Vector2Field(prop.name, new Vector2(a[0], a[1]));
```
Passing the creation of `VectorXField` or `VectorXIntField` to another handler that accepts a parameter X will probably allow us to remove a lot of duplicate code between `DrawArrayFieldsFloat()` and `DrawArrayFieldsInt()`

schinkowski replied:
Done now

sindharta commented on Editor/Scripts/MeshSyncServerInspectorUtils.cs:
```diff
@@ -8,6 +8,231 @@ namespace Unity.MeshSync.Editor
 {
     public static class MeshSyncServerInspectorUtils
```
Hm, this might have got merged by accident ? but it should be made to `internal`

schinkowski replied:
This also allows users to extend the MeshSync functionality.
I need it public for the tooling I've built on top of it for Bay's team.

sindharta replied:
If it's a public API, the question is whether we are ready to support/maintain it for public users, not just for a particular team, because changing/removing any public API 
1. may have some repercussions on the release procedure, such as having to raise minor/major version versus just patch version.
2. may break users' project

Plus, every public API requires scripting API documentations, meaning more maintenance costs.

If you just need to support this for a particular team, I would suggest using [InternalsVisibleToAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute?view=net-6.0).

schinkowski replied:
The assembly names are  just standard unity assembly names like "Assembly-CSharp" or "Assembly-CSharp-Editor", so I can't really use that.
I guess I'll just copy the code over to the other project.

schinkowski replied:
Ok, I think I figured it out, it's internal now.

sindharta commented on Editor/Scripts/MeshSyncServerInspectorUtils.cs:
```diff
@@ -8,6 +8,231 @@ namespace Unity.MeshSync.Editor
+
+                if (prop.path != lastPathDrawn)
+                {
+                    bool wasExpanded = foldouts.Contains(prop.path);
```
Perhaps using `HashSet` for foldouts is more appropriate in this case ?

schinkowski replied:
Done

sindharta commented on Editor/Scripts/MeshSyncServerInspectorUtils.cs:
```diff
@@ -8,6 +8,231 @@ namespace Unity.MeshSync.Editor
+
+            foreach (var prop in props)
+            {
+                if (parentNameFilters != null && Array.IndexOf(parentNameFilters, prop.path) == -1)
```
Can we put `parentNameFilters` null check outside the loop ?

schinkowski replied:
No, this loop goes through all properties and draws the ones that are in the parent list or draws them all if the list is null.

sindharta commented on Editor/Scripts/MeshSyncServerInspectorUtils.cs:
```diff
@@ -8,6 +8,231 @@ namespace Unity.MeshSync.Editor
+                        int max = (int)prop.max;
+
+                        // Need to be careful with overflow here:
+                        if (prop.max == int.MaxValue)
```
`Mathf.Min()` can make the code more concise ?

schinkowski replied:
Done

sindharta commented on Plugin~/Src/MeshSyncTest/TestMeshSyncClient.cpp:
```diff
@@ -18,8 +19,101 @@
+            assert(prop->type == ms::PropertyInfo::Type::String && "type is not correct");
+            scene->propertyInfos.push_back(prop);
+        }
+    }
```
Please test IntArray and FloatArray as well.

schinkowski replied:
Done

sindharta commented on Plugin~/Src/MeshSync/Include/MeshSync/msProtocol.h:
```diff
@@ -214,4 +219,37 @@ using super = Message;
 
+
+/// <summary>
+/// Message that doesn't time out and is only replied to by the server
```
Maybe we can add something along the line of
"A mechanism to initiate the sending of data from server to client"

schinkowski replied:
Done

sindharta commented on Plugin~/Src/MeshSync/Include/MeshSync/msProtocol.h:
```diff
@@ -214,4 +219,37 @@ using super = Message;
+
+class ServerInitiatedMessageResponse {
+public:
+    std::vector<PropertyInfo> properties;
```
Like what mentioned in another conversation, it seems that `ServerInitiatedMessageResponse` is not "generic" and should be renamed to reflect what it actually does: ServerLiveEdit, or Server2ClientSync, or something along this line.

sindharta commented on Plugin~/Src/MeshSync/SceneGraph/msCurve.cpp:
```diff
@@ -0,0 +1,113 @@
+		auto ret = super::hash();
+
+		for (auto spline : splines) {
+			ret += csum(spline->closed);
```
I think we should implement `vhash` for boolean, rather than using `csum` here

schinkowski replied:
Would the vhash not just return 0 or 1 anyway, same as csum?

sindharta replied:
in the current implementation, yes probably, but the intention here is to call a vhash() method right ?
If we decide to change the implementation later, there is little chance that we'll get back to this code to change it to vhash.

sindharta replied:
Bumping up

schinkowski replied:
Done

sindharta commented on Plugin~/Src/MeshSync/SceneGraph/msEntityConverter.cpp:
```diff
@@ -258,6 +285,18 @@ void FlipYZ_ZUpCorrector::convertMesh(Mesh &e)
+void FlipYZ_ZUpCorrector::convertCurve(Curve& e) {
+    convertTransform(e);
+
+    auto convert = [this](auto& v) { return flip_z(swap_yz(v)); };
```
It's probably better to turn this into a shared function in the class, rather than defining it in every convertX() functions.
But I think it's ok to leave this as it is for now.

sindharta commented on Plugin~/Src/MeshSync/SceneGraph/msPropertyInfo.cpp:
```diff
@@ -0,0 +1,110 @@
+		EachMember(msRead);
+	}
+
+#undef EachMember(F)
```
the `(F)` causes the following warning
```
unexpected tokens following preprocessor directive - expected a newline
```

schinkowski replied:
Fixed

sindharta commented on Plugin~/Src/MeshSync/SceneGraph/msPropertyInfo.cpp:
```diff
@@ -0,0 +1,110 @@
+		return ret;
+	}
+
+	uint64_t PropertyInfo::hash()
```
Is this used ? 
It seems that `m_hash` is never set

schinkowski replied:
I've removed this now, turns out we don't need it.

sindharta commented on Plugin~/Src/MeshSync/SceneGraph/msPropertyInfo.cpp:
```diff
@@ -0,0 +1,110 @@
+	static inline void set_impl(SharedVector<char>& dst, const T* v, size_t n)
+	{
+		dst.resize_discard(sizeof(T) * n);
+		memcpy(dst.data(), v, dst.size());
```
isn't `dst.size()` expected to be the same as `sizeof(T) * n` above ?
If so, it's probably clearer if we use the same value directly.

schinkowski replied:
Fixed

Commit:
Encapsulating field.

sindharta commented on Plugin~/Src/MeshSync/msServer.cpp:
```diff
@@ -608,6 +619,56 @@ void Server::recvPoll(HTTPServerRequest& request, HTTPServerResponse& response)
+    m_pending_entities.clear();
+    
+    if (m_syncRequested) {
+        reqResponse.message = REQUEST_SYNC;
```
Clarifying, `REQUEST_SYNC` requests the client to send data back to Unity after it receives `ServerInitiatedMessageResponse` ?

schinkowski replied:
Yes

sindharta commented on Runtime/Unity.MeshSync.asmdef:
```diff
@@ -35,6 +39,16 @@
             "define": "AT_USE_URP"
+        },
+        {
+            "name": "com.unity.splines",
```
Can you clarify what function is not available if we are not adding splines/probuilder in the project ?

We should probably write this in the documentation as well.

schinkowski replied:
This is in BidirectionalSupport.md.
Splines are not shown in Unity without the splines package.
Meshes can not be created as ProBuilder meshes without the ProBuilder package.

sindharta commented on Runtime/Scripts/msNetworkAPI.cs:
```diff
@@ -101,6 +106,43 @@ internal struct Server {
+        
+#if AT_USE_SPLINES
+    [DllImport(Lib.name)]
+    static extern void msServerSendTransform(IntPtr self, string path, float3 position, float3 scale, float3 rotation);  
```
A stupid question, but I assume making this into Vector3, and therefore eliminating the need of using `AT_USE_SPLINES` wouldn't make it work ?

schinkowski replied:
msServerSendTransform isn't implemented yet, there is a ticket for that. We can decide what to use then.
It's better to use float3 for msServerSendCurve to avoid a lot of conversions.

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,300 @@
+
+        protected override void SplineChanged(Spline spline, int arg2, SplineModification arg3)
+        {
+            // TODO: Get the path here:
```
Is this TODO still relevant ?

schinkowski replied:
Yes because the current API only gives us the unity spline object that has changed and we have no good way to map that back to the spline that changed so we send all splines back. We could optimise that to only send the changed spline back if we can get the object path of it somehow. I've asked the teamm to add something like a custom field to attach user data to these data structures.

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,300 @@
+#if UNITY_EDITOR
+        public override InstanceHandlingType InstanceHandling
+        {
+            get => base.InstanceHandling; set
```
`get` and `set` in the same line ? 

I suppose we can reformat this later with the other source code.

schinkowski replied:
Fixed

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,300 @@
+
+        void SendTransform(string path, EntityRecord entity)
+        {
+            // TODO: Implement this.
```
This comment is still relevant, isn't it ?

schinkowski replied:
Yes

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,300 @@
+#endif
+
+        void OnRecvPropertyRequest()
+        {
```
Does nothing when receiving `NetworkMessageType.RequestServerInitiatedMessage` ?
That shouldn't be the case, should it ?

schinkowski replied:
Right now it doesn't do anything but in the future it might be nice to add a flag or something to indicate the bidirectional support is active.

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1775,49 +1786,199 @@ EntityRecord UpdateInstancedEntity(TransformData data)
+            // Look for prefab or create one:
+            for (int prefabListIdx = 0; prefabListIdx < m_prefabList.Count; prefabListIdx++) {
+                var prefab = m_prefabList[prefabListIdx];
+                if (prefab.name == data.path) {
```
Assuming that the path is unique, we can use `SerializedDictionary` for `m_prefabList`.

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1775,49 +1786,199 @@ EntityRecord UpdateInstancedEntity(TransformData data)
+            objTransform.position   = newMat.MultiplyPoint(Vector3.zero);
+
+            // Calculate rotation here to avoid gimbal lock issue:
+            Vector3 forward;
```
same as `newMat.GetColumn(2)` or its `GetRow()` counterpart perhaps?

schinkowski replied:
Probably but this is faster

sindharta replied:
Really ? This is fine, but I'd like to know how you measured it.
In my understanding, GetColumn() or GetRow() should be faster because it's a vector operation, and because it's a native API, rather than C# script.

sindharta replied:
Bumping up

schinkowski replied:
The methods are implemented in C# and call the properties that I'm calling directly here so they are slower.

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1775,49 +1786,199 @@ EntityRecord UpdateInstancedEntity(TransformData data)
+                    }
+                    else {
+#if UNITY_EDITOR
+                        instancedCopy = (GameObject)PrefabUtility.InstantiatePrefab(instanceObjectOriginal,
```
Can't we do the same thing as runtime in the editor ?

schinkowski replied:
The PrefabUtility does not exist outside the editor. PrefabUtility.InstantiatePrefab creates objects like GameObject.Instantiate but keeps the prefab link intact which is what we want in the editor but for runtime we don't need that.

sindharta commented on Runtime/Scripts/BaseMeshSync.Modifiers.cs:
```diff
@@ -0,0 +1,544 @@
+            }
+
+            if (valueToSerialize == null)
+                return "n";
```
might be better to store these string constants into a const string, especially since they are also used in `DeserializeString()` below

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/BaseMeshSync.Modifiers.cs:
```diff
@@ -0,0 +1,544 @@
+                {
+                    case PropertyInfoDataType.IntArray:
+                        {
+                            var newValueAsArray = new int[arrayLength];
```
we can probably remove duplicate code between the handling of IntArray and FloatArray by creating a generic function

schinkowski replied:
I tried that but the casting causes issues because I cannot enforce that the type is an int or float and Math.Clamp only takes those. I tried passing in the clamp function but then it doesn't compile because the floats from the vectors cannot cast to T.

sindharta commented on Runtime/Scripts/BaseMeshSync.Modifiers.cs:
```diff
@@ -0,0 +1,544 @@
+
+                                value = newValueAsArray;
+                            }
+                            else if (value is Vector4 newValueAsVector4)
```
Should we handle the VectorXInt as well ?

schinkowski replied:
Yeah, done

sindharta commented on Runtime/Scripts/BaseMeshSync.Modifiers.cs:
```diff
@@ -0,0 +1,544 @@
+
+            if (useNewValues && newValue != null)
+            {
+                valueToSerialize = newValue;
```
What is the difference between `propertyValue` and `newValue` ?

Seems like `propertyValue` is the value obtained after converting from saved `propertyValueSerialized` 
while `newValue` is the "new" value after editing in Unity ?

Why do we need to keep both rather than just overwriting `propertyValue` directly ?

schinkowski replied:
propertyValue is what's received from blender.
newValue is what is set in Unity by the user and will be sent to blender.
We need both because we can receive syncs with new propertyValues from blender before Unity sent off the new values, in which case they would get overwritten and lost. There is logic in UpdateProperties() to ensure the newValue is kept intact when new properties are received from blender. This fixes race conditions between the client and server when a lot of properties are sent back and forth while the user is moving sliders.

sindharta replied:
I see. Can you add the following comments in the code ?

> propertyValue is what's received from blender.
> newValue is what is set in Unity by the user and will be sent to blender.

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/BaseMeshSync.Modifiers.cs:
```diff
@@ -0,0 +1,544 @@
+
+                var numSplines = data.numSplines;
+
+                Spline.Changed -= SplineChanged;
```
Since this is a static callback, can't we just set it once ?

schinkowski replied:
I need to remove it while I'm rebuilding the splines from blender so it doesn't fire at that point.

sindharta commented on Editor/Scripts/MeshSyncServerInspector.cs:
```diff
@@ -86,23 +153,58 @@ public void DrawServerSettings(MeshSyncServer t)
+        t.foldInstanceSettings = EditorGUILayout.Foldout(t.foldInstanceSettings, "Instances", true, style);
+        if (t.foldInstanceSettings)
+        {
+            t.InstanceHandling = (BaseMeshSync.InstanceHandlingType)EditorGUILayout.EnumPopup("Instance handling", t.InstanceHandling);
```
I am using the test code to get test meshes, and when I change this InstanceHandling variable, my meshes are gone, even though they aren't geometry nodes.
Is this expected ?

![changing_instance_handling](https://user-images.githubusercontent.com/7379311/177310472-8a54fecf-1900-4c69-a2d8-ed018c59ec17.gif)

schinkowski replied:
Changing the instance handling method clears the objects and needs a full sync from blender to rebuild them

Commit:
[skip ci] Adding documenation

Commit:
[skip ci] documenation updates

Commit:
Refactoring

Commit:
Removing unused code.

Commit:
Compile warning fix

Commit:
Renaming class

Commit:
Using map instead of vector.

Commit:
[skip ci] updating comment

Commit:
Extracting string constants

Commit:
Adding more vector type support.

Commit:
Adding tests to test undoing entity converters when sending data back.

Commit:
Renaming stuff to be more clear.

Commit:
Format fix

Commit:
Removing not needed code.

sindharta commented on Plugin~/Src/MeshSync/Include/MeshSync/msProtocol.h:
```diff
@@ -31,6 +35,7 @@ class Message
         Screenshot,
         Query,
         Response,
+        RequestServerInitiatedMessage
```
Change to RequestServerLiveEdit ?

schinkowski replied:
Done

sindharta commented on Plugin~/Src/MeshSync/Include/MeshSync/msServer.h:
```diff
@@ -87,6 +90,12 @@ class Server {
     void recvText(Poco::Net::HTTPServerRequest& request, Poco::Net::HTTPServerResponse& response);
     void recvScreenshot(Poco::Net::HTTPServerRequest& request, Poco::Net::HTTPServerResponse& response);
     void recvPoll(Poco::Net::HTTPServerRequest& request, Poco::Net::HTTPServerResponse& response);
+    void recvServerInitiatedRequest(Poco::Net::HTTPServerRequest& request, Poco::Net::HTTPServerResponse& response);
```
`recvServerLiveEditRequest` ?

schinkowski replied:
Done

sindharta commented on Plugin~/Src/MeshSync/SceneGraph/msScene.cpp:
```diff
@@ -231,27 +233,62 @@ void Scene::sanitizeObjectName(std::string& name)
+        else if (cv.zup_correction_mode == ZUpCorrectionMode::RotateX) {
+            if (invert) {
+                // TODO: Could write a new RotateX_ZUpCorrector that rotates back, for now just rotate 7 more times to get back to original:
+                for (int i = 0; i < 7; i++)
```
So, are we really sure that this inverts the conversion ?
I see that you have written some unit tests, but perhaps we can share the constant (7), to make it clear ?

schinkowski replied:
I'd prefer to create new converters to undo the action and create tests for them. This was supposed to be a temporary way of doing it until I can refactor it.

sindharta commented on Plugin~/Src/MeshSync/SceneGraph/msScene.cpp:
```diff
@@ -231,27 +233,62 @@ void Scene::sanitizeObjectName(std::string& name)
+        if (cv.zup_correction_mode == ZUpCorrectionMode::FlipYZ) {
+            if (invert) {
+                // TODO: Could write a new FlipYZ_ZUpCorrector that flips back, for now just rotate 3 more times to get back to original:
+                for (int i = 0; i < 3; i++)
```
I see that you have written some unit tests, but perhaps we can share the constant (3), to make it clear ?

schinkowski replied:
I'd prefer to create new converters to undo the action and create tests for them. This was supposed to be a temporary way of doing it until I can refactor it.

sindharta replied:
I see, but this "temporary" code may stay there for quite a while until you have time to refactor it, and there's no way to connect the test to the actual code if we are not using the same constant.

schinkowski replied:
Ok done

sindharta commented on Plugin~/Src/MeshSyncTest/TestMeshSyncEntityConverters.cpp:
```diff
@@ -0,0 +1,86 @@
+        converter->convertTransform(*transform);
+    }
+
+    assert(transform->position[0] == 1 &&
```
Can we store the original values first rather than putting numeric values directly in assert ?

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/MeshSyncServer.cs:
```diff
@@ -485,16 +494,56 @@ public void StartServer()
+        StopServer();
+    }
+
+    ~MeshSyncServer()
```
Putting the following code `in OnDestroy()` should be enough. 
Do we really need to define a destructor and write the same code ?
If we do, it's probably better to put them into one function to reduce duplicate code, but I'd prefer to remove the destructor if possible.

schinkowski replied:
I've made it IDisposable now.

Commit:
Commenting code

Commit:
Refactoring probuilder editor code out to separate class.

Commit:
Renaming method to be more clear.

Commit:
Renaming message type

Commit:
Refactoring tests

Commit:
Using dispose method to clean up.

Commit:
Using hashset instead of list.

Commit:
Refactoring

sindharta commented on Plugin~/Src/MeshSyncTest/TestMeshSyncEntityConverters.cpp:
```diff
@@ -0,0 +1,95 @@
+    // 4
+    converter->convertTransform(*transform);
+
+    assert(transform->position[0] == startingPosition[0] &&
```
There's a lot of duplicates here.
Isn't it possible or can't we make an operator that compares float2/3/4 ?

schinkowski replied:
I've refactored this a bit now.

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,303 @@
+            CurrentPropertiesState = PropertiesState.None;
+        }
+
+        public bool IsConnectionToBlenderActive()
```
Is this used ?
And at this level, it's probably better to abstract out the DCC name, like `IsConnectionToDCCActive` or something similar.

schinkowski replied:
I've renamed it to be more clear. It's used by the tooling built on top of meshsync.

Commit:
Using dictionary to store prefabs

Commit:
Making MeshSyncServerInspectorUtils internal

Commit:
Renaming to be more clear.

Commit:
Refactoring tests

Commit:
Extracting constants

Commit:
Updating comments

Commit:
Adding tests for array types.

Commit:
Format fix

Commit:
Increasing timeout

Commit:
Refactoring code to make it more complicated.

Commit:
Fix for null ref

sindharta commented:
Four conversations left !
I can help with rebuilding the linux/mac bundle later today.

Commit:
Adding vhash for bools.

Commit:
Renaming method to be more clear.

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -2221,6 +2381,25 @@ bool IsSubfolder(string parentPath, string childPath)
         return false;
     }
 
+    public virtual void ClearInstancePrefabs() {
```
BaseMeshSync is a public class, so this becomes a public API.
Is this expected ?
And do we need to make it virtual ?

"private protected" is an alternative.

schinkowski replied:
I've made it internal.

sindharta-tanuwijaya committed:
update mac intel bundle
[skip ci]

sindharta committed:
Update Linux so file

sindharta-tanuwijaya committed:
Update mac silicon bundle

Commit:
Making method internal

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,306 @@
+{
+    partial class MeshSyncServer
+    {
+        public enum PropertiesState
```
Failed to notice this before. Why do we need to make this public ?
And if so, we would need the API documentation too.

schinkowski replied:
I made it internal now.

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,306 @@
+        Dictionary<EntityRecord, Matrix4x4> entityTransforms = new Dictionary<EntityRecord, Matrix4x4>();
+
+#if UNITY_EDITOR
+        public override InstanceHandlingType InstanceHandling
```
Failed to notice this before. Why do we need to make this public ?
And if so, we would need the API documentation too.

schinkowski replied:
I made it internal now.

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,306 @@
+        public override InstanceHandlingType InstanceHandling
+        {
+            get => base.InstanceHandling; 
+            set
```
If we just need to override the set, wouldn't it better to convert this into a common getter and a virtual setter methods ?

schinkowski replied:
We can't have virtual setters only, the whole property needs to be virtual.

sindharta replied:
Yeah, I meant, what about converting this into two functions GetInstanceHandling() and SetInstanceHandling().
Anyway, I don't think this is a blocker, so I am going to merge this PR if there is no other conversation left.

schinkowski replied:
Personally I'd prefer properties but don't mind either way.

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,306 @@
+        }
+#endif
+
+        public int InstanceCount
```
Failed to notice this before. Why do we need to make this public ?
And if so, we would need the API documentation too.

schinkowski replied:
I made it internal

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,306 @@
+            }
+        }
+
+        public PropertiesState CurrentPropertiesState { get; private set; }
```
Failed to notice this before. Why do we need to make this public ?
And if so, we would need the API documentation too.

schinkowski replied:
I made it internal

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,306 @@
+        }
+
+        public PropertiesState CurrentPropertiesState { get; private set; }
+        public void ResetPropertiesState()
```
Failed to notice this before. Why do we need to make this public ?
And if so, we would need the API documentation too.

schinkowski replied:
I made it internal

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,306 @@
+
+        HashSet<ProBuilderMesh> changedMeshes = new HashSet<ProBuilderMesh>();
+        
+        public override bool UseProBuilder
```
Failed to notice this before. Why do we need to make this public ?
And if so, we would need the API documentation too.

schinkowski replied:
I made it internal

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,306 @@
+        public override bool UseProBuilder
+        {
+            get => base.UseProBuilder;
+            set
```
Failed to notice this before. Why do we need to make this public ?
And if so, we would need the API documentation too.

schinkowski replied:
I made it internal

sindharta commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -0,0 +1,306 @@
+
+
+#if AT_USE_PROBUILDER && UNITY_EDITOR
+        public void MeshChanged(ProBuilderMesh mesh) {
```
Failed to notice this before. Why do we need to make this public ?
And if so, we would need the API documentation too.

schinkowski replied:
I made it internal

Commit:
Changing access to internal

Commit:
Changing access to internal.

Commit:
Changing access to internal.

Commit:
Adding test

sindharta commented on Runtime/Scripts/BaseMeshSync.Modifiers.cs:
```diff
@@ -0,0 +1,528 @@
+    // Partial class for now to make merging code easier later.
+    [Serializable]
+    partial class BaseMeshSync {
+        public enum InstanceHandlingType {
```
Failed to notice this again...
Can you make this to internal ?

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/msCoreAPI.cs:
```diff
@@ -2950,8 +3034,74 @@ public Matrix4x4[] transforms
+
+#region PropertyInfo
+
+    public enum PropertyInfoDataType
```
Failed to notice this again...
Can you make this to internal ?

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/msCoreAPI.cs:
```diff
@@ -2950,8 +3034,74 @@ public Matrix4x4[] transforms
+        FloatArray,
+        String
+    };
+    public enum PropertyInfoDataSourceType {
```
Failed to notice this again...
Can you make this to internal ?

schinkowski replied:
Done

Commit:
Changing access to internal
